lib/src/presentation/widget/stream.dart

```diff
@@ -0,0 +1,93 @@
+import 'package:flutter/material.dart';
+import 'page_view.dart';
+
+import '../../core/util/states.dart';
+import '../../core/util/constants.dart';
+import '../view/about_us.dart';
+import 'error_message.dart';
+import 'grid_view.dart';
+import 'list_view.dart';
+
+class CustomStreamBuilder extends StatefulWidget {
+  final Stream<DataState<dynamic>>? data;
+  final PageController pageController;
+  final String fatherWidget;
+
+  const CustomStreamBuilder({
+    super.key,
+    required this.data,
+    required this.pageController,
+    required this.fatherWidget,
+  });
+
+  final String loadingStateKey = 'loading key';
+  final String errorStateKey = 'error key';
+  final String emptyStateKey = 'empty key';
+  final String listViewKey = 'list_view';
+  final String gridViewKey = 'grid_view';
+  final String pageViewKey = 'page_view';
+
+  @override
+  State<CustomStreamBuilder> createState() => _CustomStreamBuilderState();
+}
+
+class _CustomStreamBuilderState extends State<CustomStreamBuilder> {
+  @override
+  Widget build(
+    BuildContext context,
+  ) {
+    return StreamBuilder(
+      initialData: DataState(
+        resultState: ResultState.loading,
+      ),
+      stream: widget.data,
+      builder: (
+        BuildContext context,
+        AsyncSnapshot snapshot,
+      ) {
+        switch (snapshot.data?.resultState) {
+          case ResultState.loading:
+            return AboutUs(
+              key: Key(
+                widget.loadingStateKey,
+              ),
+              text: ConstantsClass.textAboutUs,
+            );
+          case ResultState.error:
+            return ErrorMessage(
+              key: Key(
+                widget.errorStateKey,
+              ),
+              snapshot.error.toString(),
+            );
+          case ResultState.empty:
+            return Text(
+              key: Key(
+                widget.emptyStateKey,
+              ),
+              ConstantsClass.emptyStateText,
+            );
+          case ResultState.success:
+            if (widget.fatherWidget == widget.listViewKey) {
+              return CustomListView(
+                movies: snapshot.data.data,
+              );
+            } else if (widget.fatherWidget == widget.gridViewKey) {
+              return CustomGridView(
+                movies: snapshot.data.data,
+              );
+            } else {
+              return CustomPageView(
+                movies: snapshot.data.data,
+                pageController: widget.pageController,
+              );
+            }
```

Comment:
Instead to do an if-else you can provide to the widget wich kind of constructor use, ex: if you are in home page you know that you are going to use a ListView, if you are on popular page you know that you are going to use a Grid, etc.

Reply:
Because whit this kind of if-else in the future if you add more screens, the if-else don't going to be a accurate solution

Reply:
solved